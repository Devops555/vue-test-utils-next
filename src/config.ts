import { GlobalMountOptions } from './types'

interface GlobalConfigOptions {
  global: GlobalMountOptions
  plugins: {
    VueWrapper: Pluggable
    DOMWrapper: Pluggable
  }
}

class Pluggable {
  installedPlugins: any
  constructor() {
    this.installedPlugins = []
  }

  install(handler, options = {}) {
    if (typeof handler !== 'function') {
      console.error('plugin.install must receive a function')
      handler = () => ({})
    }
    this.installedPlugins.push({ handler, options })
  }

  extend(instance) {
    const invokeSetup = (plugin) => plugin.handler(instance) // invoke the setup method passed to install
    const bindProperty = ([property, value]: [string, any]) => {
      instance[property] =
        typeof value === 'function' ? value.bind(instance) : value
    }
    const addAllPropertiesFromSetup = (setupResult) => {
      setupResult = typeof setupResult === 'object' ? setupResult : {}
      Object.entries(setupResult).forEach(bindProperty)
    }

    this.installedPlugins.map(invokeSetup).forEach(addAllPropertiesFromSetup)
  }

  /** For testing */
  reset() {
    this.installedPlugins = []
  }
}

export const config: GlobalConfigOptions = {
  global: {},
  plugins: {
    VueWrapper: new Pluggable(),
    DOMWrapper: new Pluggable()
  }
}
