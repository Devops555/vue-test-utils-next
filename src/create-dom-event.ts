import * as eventTypes from 'dom-event-types'

interface TriggerOptions {
  code?: String
  key?: String
  keyCode?: Number
  [custom: string]: any
}

interface EventParams {
  eventType: string
  modifier: string
  meta: any
  options?: TriggerOptions
}

export const keyCodesByKeyName = {
  backspace: 8,
  tab: 9,
  enter: 13,
  esc: 27,
  space: 32,
  pageup: 33,
  pagedown: 34,
  end: 35,
  home: 36,
  left: 37,
  up: 38,
  right: 39,
  down: 40,
  insert: 45,
  delete: 46
}

function getEventProperties(eventParams: EventParams) {
  const { modifier, meta, options } = eventParams
  const keyCode =
    keyCodesByKeyName[modifier] ||
    (options && (options.keyCode || options.code))

  return {
    ...options, // What the user passed in as the second argument to #trigger
    bubbles: meta.bubbles,
    meta: meta.cancelable,
    // Any derived options should go here
    keyCode,
    code: keyCode
  }
}

function createEvent(eventParams: EventParams) {
  const { eventType, meta } = eventParams
  const metaEventInterface = window[meta.eventInterface]

  const SupportedEventInterface =
    typeof metaEventInterface === 'function' ? metaEventInterface : window.Event

  const eventProperties = getEventProperties(eventParams)

  const event = new SupportedEventInterface(
    eventType,
    // event properties can only be added when the event is instantiated
    // custom properties must be added after the event has been instantiated
    eventProperties
  )

  return event
}

function createDOMEvent(eventString: String, options?: TriggerOptions) {
  const [eventType, modifier] = eventString.split('.')
  const meta = eventTypes[eventType] || {
    eventInterface: 'Event',
    cancelable: true,
    bubbles: true
  }

  const eventParams: EventParams = { eventType, modifier, meta, options }
  const event: Event = createEvent(eventParams)
  const eventPrototype = Object.getPrototypeOf(event)

  options &&
    Object.keys(options).forEach((key) => {
      const propertyDescriptor = Object.getOwnPropertyDescriptor(
        eventPrototype,
        key
      )
      const canSetProperty = !(
        propertyDescriptor && propertyDescriptor.set === undefined
      )
      if (canSetProperty) {
        event[key] = options[key]
      }
    })
  return event
}

export { TriggerOptions, createDOMEvent }
