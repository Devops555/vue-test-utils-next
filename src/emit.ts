import { setDevtoolsHook, devtools } from 'vue'

const enum DevtoolsHooks {
  COMPONENT_EMIT = 'component:emit'
}

let events: Record<string, unknown[]>

export function emitted<T = unknown>(
  eventName?: string
): T[] | Record<string, T[]> {
  if (eventName) {
    const emitted = (events as Record<string, T[]>)[eventName]
    return emitted
  }

  return events as Record<string, T[]>
}

export const attachEmitListener = () => {
  events = {}
  // use devtools to capture this "emit"
  setDevtoolsHook(createDevTools(events))
}

function createDevTools(events): any {
  const devTools: Partial<typeof devtools> = {
    emit(eventType, ...payload) {
      if (eventType !== DevtoolsHooks.COMPONENT_EMIT) return

      // The first argument is root component
      // The second argument is  vm
      // The third argument is event
      // The fourth argument is args of event
      recordEvent(events, payload[2], payload[3])
    }
  }

  return devTools
}

function recordEvent(events, event, args) {
  // Record the event message sent by the emit
  events[event]
    ? (events[event] = [...events[event], [...args]])
    : (events[event] = [[...args]])
}
