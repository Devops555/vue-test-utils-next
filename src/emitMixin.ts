import { getCurrentInstance } from 'vue'

export const attachEmitListener = () => {
  return {
    beforeCreate() {
      let events: Record<string, unknown[]> = {}
      this.__emitted = events
      const originalEmit = getCurrentInstance().emit
      getCurrentInstance().emit = (event: string, ...args: unknown[]) => {
        events[event]
          ? (events[event] = [...events[event], [...args]])
          : (events[event] = [[...args]])
        originalEmit(event, ...args)
        return [event, ...args]
      }
    }
  }
}
