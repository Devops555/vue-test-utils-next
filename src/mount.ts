import {
  h,
  createApp,
  VNode,
  defineComponent,
  VNodeNormalizedChildren,
  transformVNodeArgs,
  reactive,
  ComponentPublicInstance,
  ComponentOptionsWithObjectProps,
  ComponentOptionsWithArrayProps,
  ComponentOptionsWithoutProps,
  ExtractPropTypes,
  Component,
  VNodeProps,
  WritableComputedOptions,
  SetupContext,
  RenderFunction,
  ComponentPropsOptions
} from 'vue'

import { config } from './config'
import { GlobalMountOptions } from './types'
import { mergeGlobalProperties, isString } from './utils'
import { createWrapper, VueWrapper } from './vue-wrapper'
import { attachEmitListener } from './emitMixin'
import { createDataMixin } from './dataMixin'
import {
  MOUNT_COMPONENT_REF,
  MOUNT_ELEMENT_ID,
  MOUNT_PARENT_NAME
} from './constants'
import { stubComponents } from './stubs'

type Slot = VNode | string | { render: Function }

interface MountingOptions<Props, Data = {}> {
  data?: () => unknown extends Data ? never : Partial<Data>
  props?: Props
  attrs?: Record<string, unknown>
  slots?: {
    default?: Slot
    [key: string]: Slot
  }
  global?: GlobalMountOptions
  attachTo?: HTMLElement | string
  shallow?: boolean
}

export type ComputedOptions = Record<
  string,
  ((ctx?: any) => any) | WritableComputedOptions<any>
>
export type ObjectEmitsOptions = Record<
  string,
  ((...args: any[]) => any) | null
>
export type EmitsOptions = ObjectEmitsOptions | string[]

// Component declared with defineComponent
export function mount<TestedComponent extends ComponentPublicInstance>(
  originalComponent: { new (): TestedComponent } & Component,
  options?: MountingOptions<TestedComponent['$props'], TestedComponent['$data']>
): VueWrapper<TestedComponent>

// Functional Component declared with (props, ctx)=> render
export function mount<Props, RawBindings = object>(
  setup: (
    props: Readonly<Props>,
    ctx: SetupContext
  ) => RawBindings | RenderFunction,

  options: MountingOptions<Props, {}>
): VueWrapper<
  ComponentPublicInstance<
    Props,
    RawBindings,
    {},
    {},
    {},
    // public props
    VNodeProps & Props
  >
>

// Component declared with no props
export function mount<
  Props = {},
  RawBindings = {},
  D = {},
  C extends ComputedOptions = {},
  M extends Record<string, Function> = {},
  E extends EmitsOptions = Record<string, any>,
  EE extends string = string
>(
  componentOptions: ComponentOptionsWithoutProps<
    Props,
    RawBindings,
    D,
    C,
    M,
    E,
    EE
  >,
  options?: MountingOptions<never, D>
): VueWrapper<
  ComponentPublicInstance<Props, RawBindings, D, C, M, E, VNodeProps & Props>
>

// Component declared with { props: [] }
export function mount<
  PropNames extends string,
  RawBindings,
  D,
  C extends ComputedOptions = {},
  M extends Record<string, Function> = {},
  E extends EmitsOptions = Record<string, any>,
  EE extends string = string,
  Props extends Readonly<{ [key in PropNames]?: any }> = Readonly<
    { [key in PropNames]?: any }
  >
>(
  componentOptions: ComponentOptionsWithArrayProps<
    PropNames,
    RawBindings,
    D,
    C,
    M,
    E,
    EE,
    Props
  >,
  options?: MountingOptions<Props, D>
): VueWrapper<ComponentPublicInstance<Props, RawBindings, D, C, M, E>>

// Component declared with { props: { ... } }
export function mount<
  // the Readonly constraint allows TS to treat the type of { required: true }
  // as constant instead of boolean.
  PropsOptions extends Readonly<ComponentPropsOptions>,
  RawBindings,
  D,
  C extends ComputedOptions = {},
  M extends Record<string, Function> = {},
  E extends EmitsOptions = Record<string, any>,
  EE extends string = string
>(
  componentOptions: ComponentOptionsWithObjectProps<
    PropsOptions,
    RawBindings,
    D,
    C,
    M,
    E,
    EE
  >,
  options?: MountingOptions<ExtractPropTypes<PropsOptions>, D>
): VueWrapper<
  ComponentPublicInstance<
    ExtractPropTypes<PropsOptions>,
    RawBindings,
    D,
    C,
    M,
    E,
    VNodeProps & ExtractPropTypes<PropsOptions, false>
  >
>

// implementation
export function mount(
  originalComponent: any,
  options?: MountingOptions<any>
): VueWrapper<any> {
  const component = { ...originalComponent }

  const el = document.createElement('div')
  el.id = MOUNT_ELEMENT_ID

  if (options?.attachTo) {
    const to = isString(options.attachTo)
      ? document.querySelector(options.attachTo)
      : options.attachTo

    if (!to) {
      throw new Error(
        `Unable to find the element matching the selector ${options.attachTo} given as the \`attachTo\` option`
      )
    }
    to.appendChild(el)
  }

  // handle any slots passed via mounting options
  const slots: VNodeNormalizedChildren =
    options?.slots &&
    Object.entries(options.slots).reduce((acc, [name, slot]) => {
      // case of an SFC getting passed
      if (typeof slot === 'object' && 'render' in slot) {
        acc[name] = slot.render
        return acc
      }

      acc[name] = () => slot
      return acc
    }, {})

  // override component data with mounting options data
  if (options?.data) {
    const dataMixin = createDataMixin(options.data())
    ;(component as any).mixins = [
      ...((component as any).mixins || []),
      dataMixin
    ]
  }

  // we define props as reactive so that way when we update them with `setProps`
  // Vue's reactivity system will cause a rerender.
  const props = reactive({
    ...options?.attrs,
    ...options?.props,
    ref: MOUNT_COMPONENT_REF
  })

  // create the wrapper component
  const Parent = defineComponent({
    name: MOUNT_PARENT_NAME,
    render() {
      return h(component, props, slots)
    }
  })

  const setProps = (newProps: Record<string, unknown>) => {
    for (const [k, v] of Object.entries(newProps)) {
      props[k] = v
    }

    return vm.$nextTick()
  }

  // create the app
  const app = createApp(Parent)

  const global = mergeGlobalProperties(config.global, options?.global)

  // global mocks mixin
  if (global?.mocks) {
    const mixin = {
      beforeCreate() {
        for (const [k, v] of Object.entries(global.mocks)) {
          this[k] = v
        }
      }
    }

    app.mixin(mixin)
  }

  // use and plugins from mounting options
  if (global?.plugins) {
    for (const use of global.plugins) app.use(use)
  }

  // use any mixins from mounting options
  if (global?.mixins) {
    for (const mixin of global.mixins) app.mixin(mixin)
  }

  if (global?.components) {
    for (const key of Object.keys(global.components))
      app.component(key, global.components[key])
  }

  if (global?.directives) {
    for (const key of Object.keys(global.directives))
      app.directive(key, global.directives[key])
  }

  // provide any values passed via provides mounting option
  if (global?.provide) {
    for (const key of Reflect.ownKeys(global.provide)) {
      // @ts-ignore: https://github.com/microsoft/TypeScript/issues/1863
      app.provide(key, global.provide[key])
    }
  }

  // add tracking for emitted events
  app.mixin(attachEmitListener())

  // stubs
  if (options?.global?.stubs || options?.shallow) {
    stubComponents(options?.global?.stubs, options?.shallow)
  } else {
    transformVNodeArgs()
  }

  // mount the app!
  const vm = app.mount(el)

  const App = vm.$refs[MOUNT_COMPONENT_REF] as ComponentPublicInstance
  return createWrapper(app, App, setProps)
}

export function shallowMount(
  originalComponent,
  options?: MountingOptions<any>
) {
  return mount(originalComponent, { ...options, shallow: true })
}
