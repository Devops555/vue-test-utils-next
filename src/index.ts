import { mount } from './mount'
import { RouterLinkStub } from './components/RouterLinkStub'
import { VueWrapper } from './vue-wrapper'
import { config } from './config'

export { mount, RouterLinkStub, VueWrapper, config }
