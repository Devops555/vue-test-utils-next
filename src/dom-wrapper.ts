import { nextTick } from 'vue'

import { WrapperAPI } from './types'
import { ErrorWrapper } from './error-wrapper'

import { TriggerOptions, createDOMEvent } from './create-dom-event'

export class DOMWrapper<ElementType extends Element> implements WrapperAPI {
  element: ElementType

  constructor(element: ElementType) {
    this.element = element
  }

  classes(className?) {
    const classes = this.element.classList

    if (className) return classes.contains(className)

    return Array.from(classes)
  }

  attributes(key?: string) {
    const attributes = this.element.attributes
    const attributeMap = {}
    for (let i = 0; i < attributes.length; i++) {
      const att = attributes.item(i)
      attributeMap[att.localName] = att.value
    }

    return key ? attributeMap[key] : attributeMap
  }

  exists() {
    return true
  }

  text() {
    return this.element.textContent?.trim()
  }

  html() {
    return this.element.outerHTML
  }

  find<K extends keyof HTMLElementTagNameMap>(
    selector: K
  ): DOMWrapper<HTMLElementTagNameMap[K]> | ErrorWrapper
  find<K extends keyof SVGElementTagNameMap>(
    selector: K
  ): DOMWrapper<SVGElementTagNameMap[K]> | ErrorWrapper
  find<T extends Element>(selector: string): DOMWrapper<T> | ErrorWrapper
  find(selector: string): DOMWrapper<Element> | ErrorWrapper {
    const result = this.element.querySelector(selector)
    if (result) {
      return new DOMWrapper(result)
    }

    return new ErrorWrapper({ selector })
  }

  get<K extends keyof HTMLElementTagNameMap>(
    selector: K
  ): DOMWrapper<HTMLElementTagNameMap[K]>
  get<K extends keyof SVGElementTagNameMap>(
    selector: K
  ): DOMWrapper<SVGElementTagNameMap[K]>
  get<T extends Element>(selector: string): DOMWrapper<T>
  get(selector: string): DOMWrapper<Element> {
    const result = this.find(selector)
    if (result instanceof ErrorWrapper) {
      throw new Error(`Unable to find ${selector} within: ${this.html()}`)
    }

    return result
  }

  findAll<K extends keyof HTMLElementTagNameMap>(
    selector: K
  ): DOMWrapper<HTMLElementTagNameMap[K]>[]
  findAll<K extends keyof SVGElementTagNameMap>(
    selector: K
  ): DOMWrapper<SVGElementTagNameMap[K]>[]
  findAll<T extends Element>(selector: string): DOMWrapper<T>[]
  findAll(selector: string): DOMWrapper<Element>[] {
    return Array.from(this.element.querySelectorAll(selector)).map(
      (x) => new DOMWrapper(x)
    )
  }

  private async setChecked(checked: boolean = true) {
    // typecast so we get type safety
    const element = (this.element as unknown) as HTMLInputElement
    const type = this.attributes().type

    if (type === 'radio' && !checked) {
      throw Error(
        `wrapper.setChecked() cannot be called with parameter false on a '<input type="radio" /> element.`
      )
    }

    // we do not want to trigger an event if the user
    // attempting set the same value twice
    // this is because in a browser setting checked = true when it is
    // already true is a no-op; no change event is triggered
    if (checked === element.checked) {
      return
    }

    element.checked = checked
    return this.trigger('change')
  }

  setValue(value?: any) {
    const element = (this.element as unknown) as HTMLInputElement
    const tagName = element.tagName
    const type = this.attributes().type

    if (tagName === 'OPTION') {
      return this.setSelected()
    } else if (tagName === 'INPUT' && type === 'checkbox') {
      return this.setChecked(value)
    } else if (tagName === 'INPUT' && type === 'radio') {
      return this.setChecked(value)
    } else if (
      tagName === 'INPUT' ||
      tagName === 'TEXTAREA' ||
      tagName === 'SELECT'
    ) {
      element.value = value

      if (tagName === 'SELECT') {
        return this.trigger('change')
      }
      this.trigger('input')
      // trigger `change` for `v-model.lazy`
      return this.trigger('change')
    } else {
      throw Error(`wrapper.setValue() cannot be called on ${tagName}`)
    }
  }

  private setSelected() {
    const element = (this.element as unknown) as HTMLOptionElement

    if (element.selected) {
      return
    }

    element.selected = true
    let parentElement = element.parentElement

    if (parentElement.tagName === 'OPTGROUP') {
      parentElement = parentElement.parentElement
    }

    return new DOMWrapper(parentElement).trigger('change')
  }

  async trigger(eventString: string, options?: TriggerOptions) {
    if (options && options['target']) {
      throw Error(
        `[vue-test-utils]: you cannot set the target value of an event. See the notes section ` +
          `of the docs for more detailsâ€”` +
          `https://vue-test-utils.vuejs.org/api/wrapper/trigger.html`
      )
    }

    const isDisabled = () => {
      const validTagsToBeDisabled = [
        'BUTTON',
        'COMMAND',
        'FIELDSET',
        'KEYGEN',
        'OPTGROUP',
        'OPTION',
        'SELECT',
        'TEXTAREA',
        'INPUT'
      ]
      const hasDisabledAttribute = this.attributes().disabled !== undefined
      const elementCanBeDisabled = validTagsToBeDisabled.includes(
        this.element.tagName
      )

      return hasDisabledAttribute && elementCanBeDisabled
    }

    if (this.element && !isDisabled()) {
      const event = createDOMEvent(eventString, options)
      this.element.dispatchEvent(event)
    }

    return nextTick
  }
}
