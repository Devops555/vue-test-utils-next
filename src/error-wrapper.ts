import { ComponentPublicInstance } from 'vue'

import { DOMWrapper } from './dom-wrapper'
import { VueWrapper } from './vue-wrapper'

interface ErrorWrapperOptions {
  selector?: string
}

export function createWrapperError(options?: ErrorWrapperOptions) {
  return new Proxy<DOMWrapper<Element>>(Object.create(null), {
    get(obj, prop) {
      switch (prop) {
        case 'exists':
          return () => false
        default:
          throw new Error(`Cannot call ${String(prop)} on an empty DOMWrapper.`)
      }
    }
  })
}

export function createVueWrapperError<T extends ComponentPublicInstance>(
  options?: ErrorWrapperOptions
) {
  return new Proxy<VueWrapper<T>>(Object.create(null), {
    get(obj, prop) {
      switch (prop) {
        case 'exists':
          return () => false
        default:
          throw new Error(`Cannot call ${String(prop)} on an empty VueWrapper.`)
      }
    }
  })
}
