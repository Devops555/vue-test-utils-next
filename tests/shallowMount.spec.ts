import { mount, shallowMount } from '../src'
import ComponentWithChildren from './components/ComponentWithChildren.vue'

describe('shallowMount', () => {
  it('stubs all components automatically using { shallow: true }', () => {
    const wrapper = mount(ComponentWithChildren, { shallow: true })
    expect(wrapper.html()).toEqual(
      '<div class="ComponentWithChildren">' +
        '<hello-stub></hello-stub>' +
        '<component-with-input-stub></component-with-input-stub>' +
        '<component-without-name-stub></component-without-name-stub>' +
        '<with-props-stub></with-props-stub>' +
        '</div>'
    )
  })

  it('stubs all components automatically using shallowMount', () => {
    const wrapper = shallowMount(ComponentWithChildren)
    expect(wrapper.html()).toEqual(
      '<div class="ComponentWithChildren">' +
        '<hello-stub></hello-stub>' +
        '<component-with-input-stub></component-with-input-stub>' +
        '<component-without-name-stub></component-without-name-stub>' +
        '<with-props-stub></with-props-stub>' +
        '</div>'
    )
  })

  it.only('stubs all components, but allows providing custom stub', () => {
    const wrapper = mount(ComponentWithChildren, {
      shallow: true,
      global: {
        stubs: {
          Hello: { template: '<div>Override</div>' }
        }
      }
    })
    expect(wrapper.html()).toEqual(
      '<div class="ComponentWithChildren">' +
        '<div>Override</div>' +
        '<component-with-input-stub></component-with-input-stub>' +
        '<component-without-name-stub></component-without-name-stub>' +
        '<with-props-stub></with-props-stub>' +
        '</div>'
    )
  })
})
