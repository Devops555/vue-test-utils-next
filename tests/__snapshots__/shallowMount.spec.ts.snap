// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shallowMount renders props for stubbed component in a snapshot 1`] = `<my-label-stub val="username"></my-label-stub>`;
